File: src/etf_modules/real_bringup/scripts/test_move_xarm6_gazebo.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import rclpy
from rclpy.node import Node
from builtin_interfaces.msg import Duration

from std_msgs.msg import String
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint
from control_msgs.action import GripperCommand as GripperCommandAction
from control_msgs.msg import GripperCommand
from rclpy.action import ActionClient
import numpy as np

class TestJointTrajectoryNode(Node):

    def __init__(self):
        super().__init__('trajectory_publisher_node')
        trajectory_topic = '/xarm6_traj_controller/joint_trajectory'
        gripper_topic = '/xarm_gripper/gripper_action'
        self.trajectory_publisher = \
            self.create_publisher(JointTrajectory, trajectory_topic, 10)

        self.gripper_client = ActionClient(self, GripperCommandAction, gripper_topic)

        self.pick = True

        timer_period = 7.0
        self.timer = self.create_timer(timer_period,
                self.timer_callback)
        self.joints = [
            'joint1',
            'joint2',
            'joint3',
            'joint4',
            'joint5',
            'joint6',
            ]

        self.gripper_joint = ['drive_joint']

    def pick_procedure(self):
        trajectory = JointTrajectory()
        trajectory.joint_names = self.joints
        joint_point1 = JointTrajectoryPoint()
        joint_point1.positions = [
            0.0,
            0.0,
            0.0,
            np.pi,
            np.pi / 2,
            0.0,
            ]
            
        joint_point1.time_from_start = Duration(sec=1)
        trajectory.points.append(joint_point1)

        joint_point2 = JointTrajectoryPoint()
        joint_point2.positions = [
            0.5,
            0.0,
            -0.6,
            np.pi,
            np.pi / 2,
            0.0,
            ]
        joint_point2.time_from_start = Duration(sec=2)
        trajectory.points.append(joint_point2)
        return trajectory

    def place_procedure(self):
        trajectory = JointTrajectory()
        trajectory.joint_names = self.joints
        joint_point3 = JointTrajectoryPoint()
        joint_point3.positions = [
            -0.5,
            0.0,
            -0.6,
            np.pi,
            np.pi / 2,
            0.0,
            ]
        joint_point3.time_from_start = Duration(sec=1)
        trajectory.points.append(joint_point3)

        joint_point4 = JointTrajectoryPoint()
        joint_point4.positions = [
            0.0,
            0.0,
            0.0,
            np.pi,
            np.pi / 2,
            0.0,
            ]

        joint_point4.time_from_start = Duration(sec=2)
        trajectory.points.append(joint_point4)
        return trajectory

    def timer_callback(self):
        if (self.pick):
            trajectory = self.pick_procedure()
            self.trajectory_publisher.publish(trajectory)
            ## GRIPPER ##########################################

            gripper_command = GripperCommandAction.Goal()
            gripper_command.command = GripperCommand(position=0.1, max_effort=5.0)
            self.gripper_client.send_goal_async(gripper_command)
            self.pick = False

        else:
            trajectory = self.place_procedure()
            self.trajectory_publisher.publish(trajectory)
            ## GRIPPER ##########################################

            gripper_command = GripperCommandAction.Goal()
            gripper_command.command = GripperCommand(position=0.8, max_effort=5.0)
            self.gripper_client.send_goal_async(gripper_command)
            self.pick = True
        


def main(args=None):
    rclpy.init(args=args)
    joint_publisher = TestJointTrajectoryNode()
    rclpy.spin(joint_publisher)
    joint_publisher.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/etf_modules/sim_bringup/scripts/test_move_xarm6_gazebo.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "trajectory_msgs/msg/joint_trajectory.hpp"
#include "trajectory_msgs/msg/joint_trajectory_point.hpp"

using namespace std::chrono_literals;

class TestPlannersNode : public rclcpp::Node
{
private:
    rclcpp::TimerBase::SharedPtr timer;
    rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr trajectory_publisher;

public:
    TestPlannersNode() : Node("trajectory_publisher_node")
    {
        trajectory_publisher = this->create_publisher<trajectory_msgs::msg::JointTrajectory>("/xarm6_traj_controller/joint_trajectory", 10);
        timer = this->create_wall_timer(5s, std::bind(&TestPlannersNode::test_planners_callback, this));
    }

private:
    void test_planners_callback()
    {
        trajectory_msgs::msg::JointTrajectory trajectory;
        trajectory.joint_names = {"joint1", "joint2", "joint3", "joint4", "joint5", "joint6"};
        
        trajectory_msgs::msg::JointTrajectoryPoint point1;
        point1.positions = {0, 0, 0, 0, 0, 0};
        point1.time_from_start.sec = 1;

        trajectory_msgs::msg::JointTrajectoryPoint point2;
        point2.positions = {1.56, 0, 0, 0, 0, 0};
        point2.time_from_start.sec = 2;

        trajectory.points.emplace_back(point1);
        trajectory.points.emplace_back(point2);

        std::cout << "Publishing trajectory ...\n";
        trajectory_publisher->publish(trajectory);
    }
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TestPlannersNode>());
    rclcpp::shutdown();
    return 0;
}